File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.cpp
int minSetSize(vector<int>& arr) {
        map<int,int>mp;
        
        for(int i=0;i<arr.size();i++)
        {
            mp[arr[i]]++;
        }
        
        priority_queue<pair<int,int>>pq;
        
        for(auto it:mp)
        {
            pq.push({it.second,it.first});
        }
        
        int size=arr.size()/2;
        int count=0;
        
        while(size>0 && !pq.empty())
        {
            if(size<pq.top().first)
            {
                size-=pq.top().first;
                    count++;
                break;
            }
            else
            {
               size-=pq.top().first;
                count++;
                pq.pop();
            }
        }
        return count;
    }

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.java
class Pair{
    int num;
    int freq;
    Pair(int i,int j){
        num=i;
        freq=j;
    }
}
class Solution {
    public int minSetSize(int[] arr) {
        PriorityQueue<Pair> pq=new PriorityQueue<>((a,b)->b.freq-a.freq);
        Map<Integer,Integer> map=new HashMap<>();
        for(int i:arr){
            map.put(i,map.getOrDefault(i,0)+1);
        }
        for(Map.Entry<Integer,Integer> entry : map.entrySet()){
            pq.add(new Pair(entry.getKey(),entry.getValue()));
        }
        
        int ans=0;
        int size=arr.length;
        int sz=arr.length;
        while(sz>size/2){
            ans++;
            Pair p=pq.remove();
            sz-=p.freq;
        }
        return ans;
    }
}